gateway-spi/src/main/java/org/apache/knox/gateway/services/ServiceType.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ */
+
+package org.apache.knox.gateway.services;
+
+public enum ServiceType {
+  ALIAS_SERVICE("AliasService"),
+  CLUSTER_CONFIGURATION_MONITOR_SERVICE("ClusterConfigurationMonitorService"),
+  CRYPTO_SERVICE("CryptoService"),
+  HOST_MAPPING_SERVICE("HostMappingService"),
+  KEYSTORE_SERVICE("KeystoreService"),
+  MASTER_SERVICE("MasterService"),
+  METRICS_SERVICE("MetricsService"),
+  REMOTE_REGISTRY_CLIENT_SERVICE("RemoteConfigRegistryClientService"),
+  SERVER_INFO_SERVICE("ServerInfoService"),
+  SERVICE_DEFINITION_REGISTRY("ServiceDefinitionRegistry"),
+  SERVICE_REGISTRY_SERVICE("ServiceRegistryService"),
+  SSL_SERVICE("SSLService"),
+  TOKEN_SERVICE("TokenService"),
+  TOPOLOGY_SERVICE("TopologyService");
+
+  public static final ServiceType[] PREFERRED_START_ORDER = {
```

Comment:
Do we need this? Shouldn't services be started in the order they are added? I think this would simplify the start/stop as well since wouldn't need to worry about checking if the service is registered.

Reply:
If services were started in the order which which they were added, then any implementation of GatewayServices would need to make sure the services were added in the correct order in the event there were dependencies.  For example, a programmer would need to  know that the Alias Service and Keystore service rely on the Master Service. And therefore the Master Service should added/started before the Alias Service and Keystore Service.  Also, the Alias Service is dependent on the Keystore service, so they need to be started in the proper order. 

Since the Gateway does not have dynamic injection, these dependencies are manually set and a programmer would be tipped off on the correct order.  However there will be no guarantee that the services will be added or started in the expected order.  

If we think that this is acceptable, then I will remove the preferred order stuff and rely on the order in which the services are added for starting and reversed for stopping.

Reply:
So in init() is where the services get added. Init requires the programmer to know how to initialize each service. I think we should do more to require each service to be setup as necessary before a dependent service. Say use constructors instead of the `.setAliasService()` as done today. It shouldn't be the responsibility to the ServiceType to know what order is required.

Reply:
ok